File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofDesign/Modifiers/Shimmer.swift
import SwiftUI

/// A modern, GPU-cheap shimmer for skeleton placeholders.
///
/// Apply `.shimmering()` to a group of skeleton shapes (`SkeletonBlock`,
/// `SkeletonCircle`, `SkeletonCapsule`, or any redacted content). A soft
/// diagonal highlight band sweeps across the *masked alpha* of the content, so
/// only the placeholder silhouettes light up — never the gaps between them.
///
/// This is the single shimmer vocabulary used by every list skeleton in the
/// app, so loading states feel consistent everywhere.
///
/// Implementation is the well-worn `AnimatableModifier` + gradient-mask
/// technique: animate one `phase` value and let SwiftUI interpolate the mask
/// stops, which is far cheaper than a per-frame `TimelineView` driving dozens
/// of placeholder cells.
public struct Shimmer: ViewModifier {
    @State private var phase: CGFloat = 0
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    private let duration: Double
    private let bounce: Bool
    private let active: Bool

    public init(duration: Double = 1.4, bounce: Bool = false, active: Bool = true) {
        self.duration = duration
        self.bounce = bounce
        self.active = active
    }

    public func body(content: Content) -> some View {
        // Reduce-motion users get the static skeleton silhouette — no sweep.
        if active && !reduceMotion {
            content
                .modifier(AnimatedMask(phase: phase))
                .animation(
                    .linear(duration: duration).repeatForever(autoreverses: bounce),
                    value: phase
                )
                .onAppear { phase = 0.8 }
        } else {
            content
        }
    }

    /// Masks the content's alpha with a moving gradient band. `animatableData`
    /// drives `phase`, so SwiftUI tweens the band across every frame.
    private struct AnimatedMask: AnimatableModifier {
        var phase: CGFloat = 0
        var animatableData: CGFloat {
            get { phase }
            set { phase = newValue }
        }
        func body(content: Content) -> some View {
            content.mask(GradientMask(phase: phase).scaleEffect(3))
        }
    }

    /// Three-stop diagonal gradient. The edge alpha dips so the band reads as a
    /// travelling glint; `scaleEffect(3)` on the parent guarantees the band
    /// fully clears the content at both ends of the sweep.
    private struct GradientMask: View {
        let phase: CGFloat
        var body: some View {
            LinearGradient(
                gradient: Gradient(stops: [
                    .init(color: .black.opacity(0.35), location: phase),
                    .init(color: .black, location: phase + 0.1),
                    .init(color: .black.opacity(0.35), location: phase + 0.2)
                ]),
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )
        }
    }
}

public extension View {
    /// Sweep a shimmer highlight across skeleton placeholders. Apply to the
    /// *inner placeholder group* of a skeleton card (not its solid chrome) so
    /// only the content silhouettes shimmer. Pass `active: false` to freeze it.
    func shimmering(active: Bool = true) -> some View {
        modifier(Shimmer(active: active))
    }
}

// MARK: - Skeleton primitives

/// A single rounded placeholder block filled with the skeleton base color.
/// Compose these inside a card-shaped container that mirrors the real card's
/// chrome, then wrap the group in `.shimmering()`.
///
/// `width == nil` means "fill the available width" (left-aligned).
public struct SkeletonBlock: View {
    private let width: CGFloat?
    private let height: CGFloat
    private let cornerRadius: CGFloat

    public init(width: CGFloat? = nil, height: CGFloat, cornerRadius: CGFloat = 6) {
        self.width = width
        self.height = height
        self.cornerRadius = cornerRadius
    }

    public var body: some View {
        RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
            .fill(Color.appSkeleton)
            .frame(width: width, height: height)
            .frame(maxWidth: width == nil ? .infinity : nil, alignment: .leading)
    }
}

/// Circular skeleton placeholder — for avatars, logos, dots.
public struct SkeletonCircle: View {
    private let diameter: CGFloat
    public init(_ diameter: CGFloat) { self.diameter = diameter }
    public var body: some View {
        Circle()
            .fill(Color.appSkeleton)
            .frame(width: diameter, height: diameter)
    }
}

/// Pill/capsule skeleton placeholder — for chips, pills, tags.
/// `width == nil` fills the available width.
public struct SkeletonCapsule: View {
    private let width: CGFloat?
    private let height: CGFloat

    public init(width: CGFloat? = nil, height: CGFloat) {
        self.width = width
        self.height = height
    }

    public var body: some View {
        Capsule(style: .continuous)
            .fill(Color.appSkeleton)
            .frame(width: width, height: height)
            .frame(maxWidth: width == nil ? .infinity : nil, alignment: .leading)
    }
}

#if DEBUG
#Preview("Shimmer") {
    VStack(alignment: .leading, spacing: 16) {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 10) {
                SkeletonCircle(40)
                VStack(alignment: .leading, spacing: 6) {
                    SkeletonBlock(width: 140, height: 12)
                    SkeletonBlock(width: 90, height: 10)
                }
                Spacer()
                SkeletonCapsule(width: 60, height: 22)
            }
        }
        .padding()
        .background(RoundedRectangle(cornerRadius: 16).fill(Color.appSurfaceElevated))
        .shimmering()
    }
    .padding()
    .background(Color.appSurface)
}
#endif

File: wagerproof-android-native/core/design/src/main/kotlin/com/wagerproof/core/design/components/Shimmer.kt
package com.wagerproof.core.design.components

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.ui.Modifier
import androidx.compose.ui.composed
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.CompositingStrategy
import androidx.compose.ui.graphics.graphicsLayer

/**
 * Skeleton shimmer — port of iOS `Modifiers/Shimmer.swift`.
 *
 * A soft diagonal highlight band sweeps across the MASKED ALPHA of the
 * content, so only the placeholder silhouettes glint — never the gaps between
 * them. Apply to the inner placeholder group of a skeleton card (not its
 * solid chrome). This is the single shimmer vocabulary for every list
 * skeleton, so loading states feel consistent everywhere.
 *
 * Same masked-alpha trick as iOS: render offscreen, then multiply alpha with
 * a travelling 3-stop gradient (edges at 35% so the band reads as a glint).
 * iOS animates `phase` 0→0.8 over 1.4 s with the mask scaled 3×; here the
 * gradient spans 3× the content diagonal for the identical clear-through.
 *
 * Pass `active = false` to freeze (reduce-motion users get the static
 * silhouette — the caller decides, matching the backgrounds' contract).
 */
fun Modifier.shimmering(
    active: Boolean = true,
    durationMillis: Int = 1400,
): Modifier = composed {
    if (!active) return@composed Modifier

    val transition = rememberInfiniteTransition(label = "shimmer")
    val phase = transition.animateFloat(
        initialValue = 0f,
        targetValue = 0.8f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis, easing = LinearEasing),
            repeatMode = RepeatMode.Restart,
        ),
        label = "shimmerPhase",
    )

    Modifier
        // Offscreen so DstIn masks the whole group's alpha, not per-draw-call.
        .graphicsLayer(compositingStrategy = CompositingStrategy.Offscreen)
        .drawWithContent {
            drawContent()
            val p = phase.value
            // 3× diagonal span mirrors iOS's `scaleEffect(3)` on the mask —
            // guarantees the band fully clears the content at both sweep ends.
            val start = Offset(-size.width, -size.height)
            val end = Offset(size.width * 2f, size.height * 2f)
            drawRect(
                brush = Brush.linearGradient(
                    colorStops = arrayOf(
                        p to Color.Black.copy(alpha = 0.35f),
                        (p + 0.1f) to Color.Black,
                        (p + 0.2f) to Color.Black.copy(alpha = 0.35f),
                    ),
                    start = start,
                    end = end,
                ),
                blendMode = BlendMode.DstIn,
            )
        }
}
